### ISSUE TITLE
About 'Discontinue dartanalyzer', could we have `--options`?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#49433]
Cf. https://github.com/dart-lang/sdk/issues/48457.

The command `dart analyze` replaces the old command `dartanalyzer` in a number of ways. However, `dart analyze` does not support the `--options` argument, which is used to enable lints and perform similar configurations that would otherwise be taken from `analysis_options.yaml` in a package.

The support for `--options` is very useful for the situation where a Dart program doesn't reside in a normal package directory structure, e.g., a script in a `bin` directory somewhere, or a bunch of different tests in a directory that aren't all using the same options.

It would be great if `dart analyze` could be enhanced such that it supports `--options` in the same way as `dartanalyzer` has always done it.

**Comment by bwilkerson** [comment]
I want to push back (gently) to understand the use cases better.

The analyzer doesn't assume that every file is in a package, instead it looks for a `.dart_tool/package_config.json` file in the parent directory chain to establish a mapping for `package:` URIs, and it looks for an `analysis_options.yaml` file in the parent directory chain to establish any overrides to the default options. Neither is required to be in a traditional package root (with a `lib` directory) nor are they required to be in the same directory; it simply takes the first such files it finds.

You mentioned two concrete use cases:
- a Dart program doesn't reside in a normal package directory structure, e.g., a script in a `bin` directory somewhere, or
- a bunch of different tests in a directory that aren't all using the same options.

The first can be handled by placing the analysis options file anywhere in the parent directory chain, such as in the `bin` directory.

The second could be handled by placing the files that share a set of options (even if that's a single file) in a subdirectory so that the analysis options file will only be applied to the tests in that subdirectory. (Which seems like it might be a useful organizing structure anyway.)

Would that satisfy your needs, or are there additional constraints that I'm not seeing?

**Comment by bwilkerson** [comment]
Thanks! That helps me better understand the use cases. To be clear, I'm not opposed to adding the support, just wanting to ensure that the support that would justify the implementation and maintenance cost. Sounds like it brings a fair bit of value to the language team, and that seems like justification enough.

**Comment by jcollins-g** [comment]
The question I might have is; does this block removing `--options` from the original tool if it is otherwise not required?    I'll leave @bwilkerson to make the call but if I don't hear otherwise, will preserve `--options` until a replacement is available as I don't think the cost of leaving it in for a bit longer is that high.

**Comment by bwilkerson** [comment]
Yes, I believe that the deprecated `dartanalyze` tool should continue to support the existing options for the time being.
